### DISCUSSION TITLE
Fill price of POST_ONLY order is not the same as price of the order

### DISCUSSION THREAD

**Discussion by dpmabo** [nautechsystems/nautilus_trader#1730] [Q&A] [answered]
hello, we recently tested some simple market maker strategy, the result shows that fill price of POST_ONLY order is from the order book rather than the price of the order. It's different from the real exchange. Is this a BUG or by design? Could you fix it or add an option to set the fill price type?

I could confirm this behavior from the code below.
```rust
pub fn simulate_fills(&self, order: &BookOrder) -> Vec<(Price, Quantity)> {
        let is_reversed = self.side == OrderSide::Buy;
        let mut fills = Vec::new();
        let mut cumulative_denominator = Quantity::zero(order.size.precision);
        let target = order.size;

        for level in self.levels.values() {
            if (is_reversed && level.price.value < order.price)
                || (!is_reversed && level.price.value > order.price)
            {
                break;
            }

            for book_order in level.orders.values() {
                let current = book_order.size;
                if cumulative_denominator + current >= target {
                    // This order has filled us, add fill and return
                    let remainder = target - cumulative_denominator;
                    if remainder.is_positive() {
                        fills.push((book_order.price, remainder));
                    }
                    return fills;
                }

                // Add this fill and continue
                fills.push((book_order.price, current));
                cumulative_denominator += current;
            }
        }

        fills
    }
}
```

**Comment by cjdsellers** [comment]
Hi @dpmabo

Thanks for taking the time to report this.

Could you confirm you're seeing this fill behavior when running a backtest, and not by separately using an order book with order objects?

The reason I asked is I checked through the matching engine code and found the logic to make sure limit prices are honored:
https://github.com/nautechsystems/nautilus_trader/blob/develop/nautilus_trader/backtest/matching_engine.pyx#L1417

We also have tests for this behavior:
https://github.com/nautechsystems/nautilus_trader/blob/develop/tests/unit_tests/backtest/test_exchange_margin.py#L2533

However, I also see a condition in the matching engine that the `book_type` has to be `L1_MBP` - otherwise the fill price is not going to be set to the orders limit price.

Could you give more information on your config and what `book_type` you're setting?

**Comment by cjdsellers** [comment]
I can confirm there was a bug in the logic here for `L2_MBP` and `L3_MBO` book types, which would not set initial fill prices to the orders limit price.

I pushed a fix to `develop` branch, if you happen to be building from source could you let me know if this fixes your issue :pray:.

**Comment by dpmabo** [comment]
hi @cjdsellers,we are using L2_MBP for testing.  I will build a dev version to verify this issue.

**Comment by dpmabo** [comment]
hi @cjdsellers, I just build nt from  source and test it, it's working but seems has bug as below.

```cython
# BUG: not update initial_fill[1...N]
initial_fill_price = initial_fill[0]
fills[0] = initial_fill
```

https://github.com/nautechsystems/nautilus_trader/blob/develop/nautilus_trader/backtest/matching_engine.pyx#L1417

**Reply by cjdsellers** [reply]
I was probably too fast to remove that Chesterton fence of the `MBP_1` limitation.

I think this will need some better tests setting up - then the bug can be fixed more comprehensively.

**Reply by cjdsellers** [reply, marked as answer]
https://github.com/nautechsystems/nautilus_trader/issues/1733

**Reply by cjdsellers** [reply]
Hi @dpmabo

This issue should now be fixed on `develop` branch from 1c9f48567a09d42400a8d85fc248d9916edd4001.

**Comment by dpmabo** [comment]
Thanks Chris,  I checkout the latest dev code and it's working now.

**Reply by cjdsellers** [reply]
Good to know, thanks for the feedback.
